### PR TITLE
Adds support for paging in menus.

### PR DESCRIPTION
**Pull request by abaire**
Allows faster menu navigation via dpad left & right

### CODE REVIEW

File: Includes/menu.hpp [hunk 1]
```diff
@@ -7,6 +7,8 @@
 #include <vector>
 #include <string>
 
+// TODO: Calculate this based off screen size?
+#define MENU_PAGE_SIZE 10
```
Comment (dracc) [inline Includes/menu.hpp:11]: Value is available as  `size_t Menu::itemsToShow` - maybe make it an argument for the `pageup()`/`pagedown()` functions?

File: Includes/menu.hpp [hunk 2]
```diff
@@ -39,6 +41,12 @@ class MenuNode : public MenuItem {
 
   void up();
   void down();
+  void pageup();
+  void pagedown();
```
Comment (dracc) [inline Includes/menu.hpp:45]: Please consider using lowerCamelCase for methods.

File: Includes/menu.hpp [hunk 3]
```diff
@@ -39,6 +41,12 @@ class MenuNode : public MenuItem {
...
+
+protected:
```
Comment (dracc) [inline Includes/menu.hpp:47]: Double `protected:` label.

File: Includes/menu.hpp [hunk 4]
```diff
@@ -80,6 +88,8 @@ class Menu {
 
   void up();
   void down();
+  void pageup();
+  void pagedown();
```
Comment (dracc) [inline Includes/menu.hpp:92]: Please consider using lowerCamelCase for methods.

**Review by dracc — CHANGES_REQUESTED** [review]
Commit message doesn't follow repository style (`<component>: <Message>`).

Great stuff!
Small changes requested.

File: Includes/menu.hpp (hunk 2 above)
Comment (abaire) [reply Includes/menu.hpp]: Done.

File: Includes/menu.hpp (hunk 3 above)
Comment (abaire) [reply Includes/menu.hpp:44]: Removed.

File: Includes/menu.hpp (hunk 4 above)
Comment (abaire) [reply Includes/menu.hpp]: Done.

File: Includes/menu.hpp (hunk 1 above)
Comment (abaire) [reply Includes/menu.hpp]: Awesome, changed!

**Review by abaire — COMMENTED** [review]
I do have a usability question around the paging behavior. Right now it wraps the same way as the single item up/down, but for short lists (< `itemsToShow`) this can lead to some skipping around that might be unexpected. I'm wondering if it should stop at the top/bottom if the number of children is <`itemsToShow`, and that makes me wonder if it should just always stop at the top/bottom for paging so it's consistent.

Do you have a strong opinion?

**Comment by dracc** [comment]
Still allowing up/down to wrap but paging stops on list end/start makes sense to me. :+1:

**Comment by abaire** [comment]
Cool, being more consistent makes sense to me as well. Fixed the behavior and rebased.

File: Includes/menu.cpp [hunk 5]
```diff
@@ -90,18 +90,50 @@ void MenuNode::addNode(std::shared_ptr<MenuItem> node) {
...
 
 void MenuNode::up() {
-  if (childNodes.size() != 0) {
-    if (selected == 0) {
-      selected = childNodes.size() - 1;
+  moveSelection(-1);
+}
+
+void MenuNode::down() {
+  moveSelection(1);
```
Comment (dracc) [inline Includes/menu.cpp:97]: Doesn't set `allowWrap`.

File: Includes/menu.cpp [hunk 6]
```diff
@@ -90,18 +90,50 @@ void MenuNode::addNode(std::shared_ptr<MenuItem> node) {
 }
 
 void MenuNode::up() {
-  if (childNodes.size() != 0) {
-    if (selected == 0) {
-      selected = childNodes.size() - 1;
+  moveSelection(-1);
```
Comment (dracc) [inline Includes/menu.cpp:93]: Doesn't set `allowWrap`.

**Comment by dracc** [comment]
Silly me didn't spot the default in the header file. Merged, thanks! :+1:

**Pull request merged**